### ISSUE TITLE
parser should recover from seeing `,` when `;` is expected

### ISSUE THREAD

**Issue by pnkfelix** [rust-lang/rust#48160]
Right now, if `rustc` is parsing a function body and sees a `,` when it expected a `;`, it will give up on parsing the rest of the current function body.

Meanwhile, later phases of the compiler still run. Which means, for example, if the remainder of that erroneous function body contains the sole references to certain imports, then you get false warnings about those imports being unused.

Here is a concrete example ([playground](https://play.rust-lang.org/?gist=e22178b3eef8977777022ea98144b439&version=beta)):

```rust
use std::iter::Iterator;

fn main() {
    struct S;
    let _x = 3,
    impl Iterator for S {
        type Item = S;
        fn next(&mut self) -> Option<S> { Some(S) }
    }
}
```

which yields the following diagnostic output on nightly and beta:

```
error: expected one of `.`, `;`, `?`, or an operator, found `,`
 --> src/main.rs:5:15
  |
5 |     let _x = 3,
  |               ^ expected one of `.`, `;`, `?`, or an operator here

warning: unused import: `std::iter::Iterator`
 --> src/main.rs:1:5
  |
1 | use std::iter::Iterator;
  |     ^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(unused_imports)] on by default

error: aborting due to previous error
```

This seems to affect the current nightly and beta, but not stable (I.e., the stable channel just presents the parse error, not the erroneous lint about the supposedly unused import.)

**Comment by pnkfelix** [comment]
(Okay, to be fair, its expecting any of `.`, `;`, or `?`, which means you cannot recover by replacing `,` with `;` in the general case of a single transposed character. But my guess is that if the user has a `,` there, it is often because they cut-and-pasted an expression for some input argument from some other line, and I'm betting `;` is a pretty good choice for recovery...)

**Comment by estebank** [comment]
The parser should do a better job of recovering one char typos. At the same time, we should be doing a better job of poisoning blocks that have had parse errors to avoid typechecking them afterwards. I've made some [small steps in this direction](https://github.com/rust-lang/rust/pull/46732) in the past, but more work is needed (and help is more than welcome).

**Comment by estebank** [comment]
CC #44767
